File: repository/extend-scader/src/main/scala/org/abstracthorizon/extend/repository/maven/MavenVersion.scala
package org.abstracthorizon.extend.repository.maven

import org.abstracthorizon.extend.repository._

object MavenVersion {

    def apply(versionString: String): MavenVersion = if (versionString != null && versionString.trim.size > 0) new MavenVersion(versionString) else null

    def toSnapshot(version: Version): MavenVersion = version match {
            case mv: MavenVersion => if (mv.isFinal) MavenVersion(mv.toString() + "-SNAPSHOT") else mv
            case v: Version => if (v.toString.endsWith("-SNAPSHOT")) MavenVersion(v.toString) else {
                val vs = version.toString
                MavenVersion(vs.substring(0, vs.length - 9))
            }
    }
}

class MavenVersion(s: String) extends SimpleVersion(s: String) {
    
    override def isFinal = !versionString.endsWith("-SNAPSHOT")

    override def toFinal = if (isFinal) this else MavenVersion(versionString.substring(0, versionString.length - 9))


}

File: repository/extend-maven/src/main/java/org/abstracthorizon/extend/repo/maven/MavenVersion.java
package org.abstracthorizon.extend.repo.maven;

import org.abstracthorizon.extend.repo.SimpleVersion;
import org.abstracthorizon.extend.repo.Version;

public class MavenVersion extends SimpleVersion {

    public MavenVersion(String versionString) {
        super(versionString);
    }
    
    public static MavenVersion apply(String versionString) {
        if (versionString != null && versionString.trim().length() > 0) {
            return new MavenVersion(versionString);
        } else {
            return null;
        }
    }
    
    public static MavenVersion toSnapshot(Version version) {
        if (version.isFinal()) {
            return new MavenVersion(version.toString() + "-SNAPSHOT");
        } else {
            if (version instanceof MavenVersion) {
                return (MavenVersion) version;
            } else {
                return new MavenVersion(version.toString());
            }
        }
    }
    
    @Override
    public boolean isFinal() {
        return !getVersionString().endsWith("-SNAPSHOT");
    }
    
    @Override
    public Version toFinal() {
        if (isFinal()) {
            return this;
        } else {
            String versionString = getVersionString();
            return new MavenVersion(versionString.substring(0, versionString.length() - 9));
        }
    }
}
